version: '3.8'

networks:
  rabbit:
    driver: bridge
  db:
    driver: bridge

services:
  calendar_app:
    build:
      context: .
      dockerfile: cmd/calendar/Dockerfile
    depends_on:
      - pg
    restart: on-failure
    ports:
      - "8888:8080"
    expose:
      - 8888
    networks:
      - db

  sender:
    build:
      context: .
      dockerfile: cmd/calendar_sender/Dockerfile
    depends_on:
      rabbitmq:
        condition: service_healthy
    networks:
      - db
      - rabbit

  scheduler:
    build:
      context: .
      dockerfile: cmd/calendar_sheduler/Dockerfile
    depends_on:
      rabbitmq:
        condition: service_healthy
    networks:
      - db
      - rabbit

  pg:
    restart: always
    image: postgres:latest
    environment:
      POSTGRES_USER: postgres
      POSTGRES_PASSWORD: postgres
      POSTGRES_DB: calendar
    ports:
      - "5435:5432"
    volumes:
      - ./internal/pg/data:/var/lib/postgresql/data
      - ./scripts/setup.sql:/docker-entrypoint-initdb.d/setup.sql
    networks:
      - db

  rabbitmq:
    image: rabbitmq:3.13.3-management
    hostname: rabbitmq
    ports:
      - "5672:5672"
      - "15675:15672"
    healthcheck:
      test: [ "CMD", "rabbitmqctl", "status" ]
      interval: 15s
      timeout: 15s
      retries: 10
    volumes:
      - ./rabbitmq:/var/lib/rabbitmq
    networks:
      - rabbit